servicetalk-data-protobuf-jersey/src/main/java/io/servicetalk/data/protobuf/jersey/ProtobufSerializerMessageBodyReaderWriter.java

```diff
@@ -0,0 +1,240 @@
+        }
+
+        setResponseBufferPublisher(bufferPublisher, requestCtxProvider.get());
+    }
+
+    private ProtobufSerializerFactory getSerializerFactory(final MediaType mediaType) {
+        final ContextResolver<ProtobufSerializerFactory> contextResolver =
+                providers.getContextResolver(ProtobufSerializerFactory.class, mediaType);
+
+        return contextResolver != null ? contextResolver.getContext(ProtobufSerializerFactory.class) :
+                ProtobufSerializerFactory.PROTOBUF;
+    }
+
+    private static Publisher<Buffer> toBufferPublisher(final InputStream is, final BufferAllocator a) {
+        return fromInputStream(is).map(a::wrap);
+    }
+
+    private static <T> Single<T> deserialize(
+            final Publisher<Buffer> bufferPublisher, final Deserializer<T> deserializer, final int contentLength,
+            final BufferAllocator allocator) {
+        return bufferPublisher
+                .collect(() -> newBufferForRequestContent(contentLength, allocator), Buffer::writeBytes)
+                .map(buf -> {
+                    try {
+                        return deserializer.deserialize(buf, allocator);
+                    } catch (final NoSuchElementException e) {
+                        throw new BadRequestException("No deserializable Protobuf content", e);
+                    } catch (final SerializationException e) {
+                        // SerializationExceptionMapper can't always tell for sure that the exception was thrown because
+                        // of bad user data: here we are deserializing user data so we can assume we fail because of it
+                        // and immediately throw the properly mapped JAX-RS exception
+                        throw new BadRequestException("Invalid Protobuf data", e);
+                    }
+                });
+    }
+
+    static Buffer newBufferForRequestContent(final int contentLength,
+                                             final BufferAllocator allocator) {
+        return contentLength == -1 ? allocator.newBuffer() : allocator.newBuffer(contentLength);
+    }
+
+    // visible for testing
+    static <T> T deserializeObject(final Publisher<Buffer> bufferPublisher, final Deserializer<T> deserializer,
+                                   final int contentLength, final BufferAllocator allocator) {
+        return awaitResult(deserialize(bufferPublisher, deserializer, contentLength, allocator).toFuture());
+    }
+
+    private static boolean isSse(ContainerRequestContext requestCtx) {
+        final ResourceMethod method = ((ExtendedUriInfo) requestCtx.getUriInfo()).getMatchedResourceMethod();
+        return method != null && method.isSse();
+    }
+
+    private static boolean isSupportedMediaType(final Class<?> type, final MediaType mediaType) {
+        return (!Publisher.class.equals(type) &&
+                (mediaType.getType().equalsIgnoreCase(APPLICATION_PROTOBUF_TYPE.getType()) &&
```

Comment:
IIUC, this method is required only because annotation on this class use `WILDCARD`. If we specify media types explicitly, we likely don't need this logic.

Reply:
We still need to match the media type against `Publisher` for `var-int` vs "other" types for regular content type.